db/commitlog/commitlog.cc

```diff
@@ -2693,6 +2693,9 @@ db::commitlog::read_log_file(sstring filename, sstring pfx, seastar::io_priority
                 if (corrupt_size > 0) {
                     throw segment_data_corruption_error("Data corruption", corrupt_size);
                 }
+            } catch (std::out_of_range&) {
+                // if we get an underflow, the file is somehow truncated/not fully created
+                p = std::make_exception_ptr(file_truncated_error{});
```

Comment:
Underflow of what?

Do you expect this to be thrown from read_header or read_chunk?

Maybe better to try/catch there (and perhaps convert the exception type)? Or check before the access?

Such a generic exception is inviting trouble if we have several places that can throw out_of_range.

Reply:
Underflow of available bytes. Could be header or chunk. We do size checking and try to reads exactly the sizes chunks/headers should be, but the behaviour of fragmented_reader::read_exactly is to silently give back less bytes if not available, so a short read will result in out_of_range a bit later. Could of course also wrap the read_exactly:s a throw this there instead. But this is further complicated by our datasource::skip not really telling us a result, so would need to make sure we use tell etc to make sure we are at expected positions. This was a shorter, more guaranteed inclusive fix.